### ISSUE TITLE
Mocked fetch does not support a call with Request as one argument

### ISSUE THREAD

**Issue by kirill-konshin** [wheresrhys/fetch-mock#156]
Mocked fetch does not support a call with Request as one argument:

```js
fetch(new Request(...))
```

Produces errors like `request.url.headers` are undefined.

I have to call it like this in my code so far:

```js
var req = new Request(...);
fetch(req.url, req)
```

**Comment by wheresrhys** [comment]
Could you tell me which methods of fetch-mock you use to build the mock?

**Comment by kirill-konshin** [comment]
The actual code is a bit more generic, but it can be narrowed to this.

```js
fetchMock.once('http://whatever', {
    body: JSON.stringify(json),
    status: 200,
    statusText: 'OK',
    headers: {'Content-Type': 'application/json'},
    sendAsJson: false
}, {
    method: 'POST'
});
```

**Comment by wheresrhys** [comment]
I can't reproduce. On line 22 of compile-route.js can you log out `Request.prototype.isPrototypeOf(url)`. It should be `true` though there are possibly scenarios (depending on your dependency tree) where the `Request` used to create your new request isn't the same as the reference that fetch-mock has

**Comment by kirill-konshin** [comment]
My bad, forgot to mention, that this occurs only when used with Karma and browserified version.

**Comment by wheresrhys** [comment]
Can you create a repo with a minimal test case and commands needed to run it and I'll investigate

**Comment by kirill-konshin** [comment]
I have sent you an email with instructions

**Comment by kirill-konshin** [comment]
I have narrowed the problem to Firefox browser, I had version 45.0, so I upgraded it to 49.2 and the problem is now gone.

**Comment by kirill-konshin** [comment]
Got a very weird behavior from Travis CI: https://travis-ci.org/ringcentral/ringcentral-js/builds/185627049.

I have removed the patch from my code:

```js
fetchMock.fetchMock = function(req) {
    return fetchMockOrig(req.url, req); // For some reasons that's needed for TravisCI
};
```

And now some versions of Node fail: 4, 0.11, 0.11, all other versions (incl. Karma) are passing. No idea why is that happening, but with patch test are OK.

The signature of `fetchMock` function is:

```js
value: function fetchMock(url, opts) {
```

And when the error occurs `url` parameter is stringified. So at least please alter the error handler to always show URL: `url.url || url` because per spec `fetch()` accepts a `Request` instance as first parameter and no second param at all.

**Comment by wheresrhys** [comment]
I'm gonna close this again. Without precise info on recreating the bug I can't really address it. Your project seems to be unblocked now though, which is good :)

**Comment by kirill-konshin** [comment]
Regarding last error: seems that somehow in older Node versions the implementation of `Request` in `fetch-mock` was different to the one used by my application. So the `normalizeRequest` was failing at the `Request.prototype.isPrototypeOf` check.

I have fixed it by accessing your `Request` implementation via `fetchMock.constructor.Request`.

I suppose the check should be weaker because no matter what implementation is supplied, things still should work if interface is preserved...

**Comment by gnapse** [comment]
I have this problem as well. `fetch` normally supports invoking it with a single argument of type [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) (according to [this documentation](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Syntax)).

But from the very definition of the mocked `fetch` implementation provided by this library [here](https://github.com/wheresrhys/fetch-mock/blob/master/src/fetch-mock.js#L83), it becomes evident that the mocked `fetch` always expect `fetch` to be called as `fetch(url, options)`. No where in that method it is considered the possibility that `url` is not a string, or that `options` could be missing. In fact, the only thing done around `option` being missing is defaulting to method `GET`. I actually realized all this when `fetch-mock` was giving me an error that there was no fallback response for a `GET` request, in an example where I was only firing a single `POST` request. The only caveat is that in my app, the requests are built as a `Request` object, which is then passed to `fetch`.

**Comment by nackjicholson** [comment]
Not sure this is related, but I think it may be.

When I send do something like:

```js
fetchMock.mock(`${baseUrl}/test`, {
  status: 404,
  headers: {'Content-Type': 'application/json'},
  body: {error: 'test.body.error.message'}
});
```

The response that actually comes back, does not have headers.

It's `Headers { _headers: {} }`

**Comment by gnapse** [comment]
I temporarily "solved" the problem in my app by wrapping the mocked `fetch` function around a version of my own that passes the parameters correctly to the original mocked function.

```javascript
import fetchMock from 'fetch-mock';

const _fetchMock = fetchMock.fetchMock;
fetchMock.fetchMock = (url, options) => {
  if (url instanceof Request && options === undefined) {
    // Function was called with a single argument of type Request
    return _fetchMock(url.url, { method: url.method });
  }
  return _fetchMock(url, options);
};
```

So far I got away with it passing only the `Request`'s `method`. YMMV, and you may need to pass additional request options if needed.

**Comment by gnapse** [comment]
@wheresrhys with all the comments in this thread after the issue was closed, can you see the problem? I really think this should be addressed, and the issue is by no means solved or inexistent. It very much exists.

**Comment by wheresrhys** [comment]
I agree, but I've asked before in this thread for somebody to provide a reduced test case. Also pull requests are welcome.

**Comment by gnapse** [comment]
I would do it, but there's a problem. I'm not sure if I could extract the `body` from the `Request` object. Everything would be easier if the entire `fetchMock` library would be developed based on `Request`, and then making a small patch to give support for when `fetch` is called with `url` and `options`.

For example, an extension of my sort of "patch" above, I'd do something around the following, in the definition of fetchMock:

```javascript
FetchMock.prototype.fetchMock = function (url, opts) {
  if (typeof opts === 'undefined' && url instanceof Request) {
    url = url.url;
    opts = url;
  }
  // ... rest of the current implementation here, untouched
}
```

Since in the above example `url` is not a string, but a `Request` object, it would (almost) work nicely as a replacement for `opts`, since `Request` instances are an object with all the usual `fetch`-supported options already as properties in it. But `body` for some reason is treaded differently, and there's no `request.body` even though you can provide a body to the request instance when creating it.

I'll put here links to the documentation for `fetch` ([here](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch)) and `Request` ([here](https://developer.mozilla.org/en-US/docs/Web/API/Request)), just in case someone wants to help so they have this in context.

**Comment by devwells** [comment]
I was having this problem and I found the cause was that I misunderstood the documentation. @wheresrhys

In my code I am doing a fetch using a request object.
`fetch(new Request(...))`

You were correct in diagnosing the issue: 
> the Request used to create your new request isn't the same as the reference that fetch-mock has

So, I tried to use the utility function you mentioned. I implemented it like so:
`fetchMock.mock('exampleurl.com', {
          status: 200,
          data: 'data'
        }).setImplementations({Request: Request});`

After looking through the code, I found out the the matching function was not being updated until after the route was pushed into the routes array. So I added a beforeEach function like this:
`  beforeEach( ()=> {
    fetchMock.setImplementations({Request: Request});
  });`

to my test suite and now everything works great! 

Just wanted to mention this in case someone is having the same issue as me :)

**Comment by tariqzafa700** [comment]
@wheresrhys 
How do we pass the same Request constructor reference? My constructor depends on username and password in body of Request. Hence it can't create an object beforehand unless my function is called.

My scenario is like this 

```
const matcherFunc = (url, opts) => {
        return (url == 'http://localhost:8080/controller/core/login' && opts && opts.method == 'POST' 
          )
      }
      let resp = {"status":200,"code":200,"message":"Success!"} 
      let payload = JSON.stringify({ username: 'user', password: 'pwd' })
      let headers = new Headers({ 'Content-Type': 'application/json' });
      let options = {method: "POST", body: payload, headers: headers, credentials: 'include'};
      fetchMock.setImplementations([{Request: Request}])
      
      fetchMock.mock(matcherFunc,
          resp,
          options
      ).catch(503);
```
I get the response
`Unmatched GET to [object Object]
`

**Comment by kopax** [comment]
I am hitting the same error with a working fetch call. But fetch-mock fail to run it:


```
    > 44 |     const res = await authClient(AUTH_LOGIN, { username, password })
```;

**Comment by wheresrhys** [comment]
Please open a new issue with more details on recreating the problem

**Comment by wheresrhys** [comment]
NOTE TO FUTURE USERS - PLEASE DO NOT REOPEN THIS ISSUE. IT RELATES TO  AMUCH EARLIER VERSIONS OF THE LIBRARY. IF YOU'RE EXPERIENCING A SIMILAR PROBLEM, PLEASE OPEN A NEW ISSUE WITH DETAILS ON HOW TO RECREATE THE PROBLEM
